common/src/main/java/io/netty/util/Recycler.java

```diff
@@ -177,478 +174,120 @@ public final boolean recycle(T o, Handle<T> handle) {
         }
 
         DefaultHandle<T> h = (DefaultHandle<T>) handle;
-        if (h.stack.parent != this) {
+        if (h.parent != this) {
             return false;
         }
 
         h.recycle(o);
         return true;
     }
 
-    final int threadLocalCapacity() {
-        return threadLocal.get().elements.length;
-    }
-
-    final int threadLocalSize() {
-        return threadLocal.get().size;
-    }
-
     protected abstract T newObject(Handle<T> handle);
 
     public interface Handle<T> {
         void recycle(T object);
     }
 
     static final class DefaultHandle<T> implements Handle<T> {
```

Comment:
should we just extend `WeakReference` here to reduce object creation ?

Reply:
The problem is that WeakReference is a fixed reference at construction time and we need WeakReference in the linked list. So the linked list alternates between the fixed WeakReference and the mutable DefaultHandle.